### PR TITLE
{jam,ftjam}: fix cross-compiling and refactor

### PR DESCRIPTION
**Pull request by impl**
###### Description of changes

This change supports cross-compiling the jam and ftjam packages. Also, it removes a bunch of duplication between the two packages and combines them into a single file.

Please note that I am employed by Perforce but I am not currently supporting these packages in an official capacity.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -5625,6 +5625,16 @@
     githubId = 510202;
     name = "Ismaël Bouya";
   };
+  impl = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:5628]: `maintainer-list.nix` should be updated in a separate commit.

Do not open a pull request, only split this patch in a separate commit

https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
Also I suggest a `rec`-less expression.

**Review by AndersonTorres — APPROVED** [review]
LGTM

**Comment by impl** [comment]
Thank you for the prompt review, @AndersonTorres! Much appreciated. :)

This one threw me for a loop at first because, like compilers, it has a two-phase build process where you have to bootstrap an initial binary and then use it to rebuild for your target. I didn't understand why I didn't have to do anything to seemingly make that work when the upstream source code didn't provide any indication it knew about cross compiling. It turned out that I had a bunch of `boot.binfmt.emulatedSystems` defined that I had forgotten about, and so it was just executing the foreign architecture for me. It's almost a purity issue, but not really, because it's still deterministic... it's more like an implicit `depsBuildBuild` dependency almost?

So I just pushed an extra commit to fix that, and I also wonder if it's something that should be more widely understood... I don't think it's possible to turn off binfmt_misc entries inside a namespace, but maybe at least there should be some warning that it's being used to execute a program inside a builder?

Anyway, I'm digressing a bit and this should be ready to go now across the board.
